File: packages/vm/wasmlib/go/wasmclient/encoder.go
package wasmclient

import "encoding/binary"

type Encoder struct{}

func fromBase58(value string, typeID int32) []byte {
	bytes := Base58Decode(value)
	if len(bytes) != int(TypeSizes[typeID]) {
		panic("invalid byte size")
	}
	return bytes
}

func (c Encoder) FromAddress(value Address) []byte {
	return fromBase58(string(value), TYPE_ADDRESS)
}

func (c Encoder) FromAgentID(value AgentID) []byte {
	return fromBase58(string(value), TYPE_AGENT_ID)
}

func (c Encoder) FromBool(value bool) []byte {
	bytes := []byte{0}
	if value {
		bytes[0] = 1
	}
	return bytes
}

func (c Encoder) FromBytes(value []byte) []byte {
	if value == nil {
		panic("missing mandatory byte array")
	}
	return value
}

func (c Encoder) FromChainID(value ChainID) []byte {
	return fromBase58(string(value), TYPE_CHAIN_ID)
}

func (c Encoder) FromColor(value Color) []byte {
	color := string(value)
	if color == COLOR_IOTA {
		color = COLOR_IOTA_BASE58
	}
	return fromBase58(color, TYPE_COLOR)
}

func (c Encoder) FromHash(value Hash) []byte {
	return fromBase58(string(value), TYPE_HASH)
}

func (c Encoder) FromHname(value Hname) []byte {
	return c.FromUint32(uint32(value))
}

func (c Encoder) FromInt8(value int8) []byte {
	return c.FromUint8(uint8(value))
}

func (c Encoder) FromInt16(value int16) []byte {
	return c.FromUint16(uint16(value))
}

func (c Encoder) FromInt32(value int32) []byte {
	return c.FromUint32(uint32(value))
}

func (c Encoder) FromInt64(value int64) []byte {
	return c.FromUint64(uint64(value))
}

func (c Encoder) FromRequestID(value RequestID) []byte {
	return fromBase58(string(value), TYPE_REQUEST_ID)
}

func (c Encoder) FromString(value string) []byte {
	return []byte(value)
}

func (c Encoder) FromUint8(value uint8) []byte {
	return []byte{value}
}

func (c Encoder) FromUint16(value uint16) []byte {
	bytes := make([]byte, 2)
	binary.LittleEndian.PutUint16(bytes, value)
	return bytes
}

func (c Encoder) FromUint32(value uint32) []byte {
	bytes := make([]byte, 4)
	binary.LittleEndian.PutUint32(bytes, value)
	return bytes
}

func (c Encoder) FromUint64(value uint64) []byte {
	bytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(bytes, value)
	return bytes
}

File: packages/vm/wasmlib/ts/wasmclient/encoder.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmclient from "./index"
import {Base58} from "./crypto";
import {Buffer} from "./buffer";

// The Arguments struct is used to gather all arguments for this smart
// contract function call and encode it into this deterministic byte array
export class Encoder {
    private fromBase58(val: string, typeID: wasmclient.Int32): Buffer {
        const bytes = Base58.decode(val);
        if (bytes.length != wasmclient.TYPE_SIZES[typeID]) {
            wasmclient.panic("invalid byte size");
        }
        return bytes;
    }

    fromAddress(val: wasmclient.AgentID): Buffer {
        return this.fromBase58(val, wasmclient.TYPE_ADDRESS);
    }

    fromAgentID(val: wasmclient.AgentID): Buffer {
        return this.fromBase58(val, wasmclient.TYPE_AGENT_ID);
    }

    fromBool(val: boolean): Buffer {
        const bytes = Buffer.alloc(1);
        if (val) {
            bytes.writeUInt8(1, 0);
        }
        return bytes;
    }

    fromBytes(val: wasmclient.Bytes): Buffer {
        return val;
    }

    fromChainID(val: wasmclient.ChainID): Buffer {
        return this.fromBase58(val, wasmclient.TYPE_CHAIN_ID);
    }

    fromColor(val: wasmclient.Color): Buffer {
        if (val == "IOTA") {
            val = "11111111111111111111111111111111";
        }
        return this.fromBase58(val, wasmclient.TYPE_COLOR);
    }

    fromHash(val: wasmclient.Hash): Buffer {
        return this.fromBase58(val, wasmclient.TYPE_HASH);
    }

    fromHname(val: wasmclient.Hname): Buffer {
        return this.fromUint32(val);
    }

    fromInt8(val: wasmclient.Int8): Buffer {
        const bytes = Buffer.alloc(1);
        bytes.writeInt8(val, 0);
        return bytes;
    }

    fromInt16(val: wasmclient.Int16): Buffer {
        const bytes = Buffer.alloc(2);
        bytes.writeInt16LE(val, 0);
        return bytes;
    }

    fromInt32(val: wasmclient.Int32): Buffer {
        const bytes = Buffer.alloc(4);
        bytes.writeInt32LE(val, 0);
        return bytes;
    }

    fromInt64(val: wasmclient.Int64): Buffer {
        const bytes = Buffer.alloc(8);
        bytes.writeBigInt64LE(val, 0);
        return bytes;
    }

    fromRequestID(val: wasmclient.RequestID): Buffer {
        return this.fromBase58(val, wasmclient.TYPE_REQUEST_ID);
    }

    fromString(val: string): Buffer {
        return Buffer.from(val);
    }

    fromUint8(val: wasmclient.Uint8): Buffer {
        const bytes = Buffer.alloc(1);
        bytes.writeUInt8(val, 0);
        return bytes;
    }

    fromUint16(val: wasmclient.Uint16): Buffer {
        const bytes = Buffer.alloc(2);
        bytes.writeUInt16LE(val, 0);
        return bytes;
    }

    fromUint32(val: wasmclient.Uint32): Buffer {
        const bytes = Buffer.alloc(4);
        bytes.writeUInt32LE(val, 0);
        return bytes;
    }

    fromUint64(val: wasmclient.Uint64): Buffer {
        const bytes = Buffer.alloc(8);
        bytes.writeBigUInt64LE(val, 0);
        return bytes;
    }
}
